staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -88,6 +83,38 @@ type delayingType struct {
 type waitFor struct {
 	data    t
 	readyAt time.Time
+	// index in the priority queue (heap)
+	index int
+}
+
+type waitForPriorityQueue []*waitFor
+
+func (pq waitForPriorityQueue) Len() int {
+	return len(pq)
+}
+func (pq waitForPriorityQueue) Less(i, j int) bool {
+	return pq[i].readyAt.Before(pq[j].readyAt)
+}
+func (pq waitForPriorityQueue) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].index = i
+	pq[j].index = j
+}
+func (pq *waitForPriorityQueue) Push(x interface{}) {
+	n := len(*pq)
+	item := x.(*waitFor)
+	item.index = n
+	*pq = append(*pq, item)
+}
+func (pq *waitForPriorityQueue) Pop() interface{} {
+	n := len(*pq)
+	item := (*pq)[n-1]
+	item.index = -1
+	*pq = (*pq)[0:(n - 1)]
+	return item
+}
+func (pq waitForPriorityQueue) Peek() interface{} {
+	return pq[0]
```

Comment:
Can you write up an example with the index field filled in?  My cursory look at push, pop, and peek seem to suggest that we peek at pq[0], pop from pq[n-1], and push onto the end.